File: Leetcode/src/com/leetcode/dynamic_programming/_926_FlipStringToMonotoneIncreasing.java
/*
 * @Date: 08/10/2021 09:51:44
 * @LastEditTime: 08/10/2021 16:12:55
 * @Description: Prefix, Suffix, DP
 */
package com.leetcode.dynamic_programming;

public class _926_FlipStringToMonotoneIncreasing {
    public int minFlipsMonoIncr(String s) {
        if (s == null || s.length() == 0) {
            return 0;
        }
        int n = s.length();
        int[] convertToOne = new int[n + 1]; // [j] -> [j, n] minimal changes to one;
        int[] convertToZero = new int[n + 1]; // [i] -> [0, i) minimal changes to zero;

        for (int i = 1, j = n - 1; j >= 0; i++, j--) {
            convertToZero[i] = convertToZero[i - 1] + (s.charAt(i - 1) == '0' ? 0 : 1);
            convertToOne[j] = convertToOne[j + 1] + (s.charAt(j) == '1' ? 0 : 1);
        }

        int res = Integer.MAX_VALUE;
        for (int i = 0; i <= n; i++) {
            res = Math.min(res, convertToZero[i] + convertToOne[i]);
        }
        return res;
    }
}

File: Leetcode/src/com/leetcode/dynamic_programming/_926_Flip_String_to_Monotone_Increasing.py
'''
Date: 08/10/2022 15:07:43
LastEditTime: 08/10/2022 15:07:44
Description: Prefix, Suffix
'''


class Solution:
    def minFlipsMonoIncr(self, s: str) -> int:
        if not s:
            return 0
        n = len(s)
        convert_to_one = [0] * (n + 1)
        convert_to_zero = [0] * (n + 1)
        # [0, i)
        for i in range(1, n + 1):
            count = (0 if s[i - 1] == '0' else 1)
            convert_to_zero[i] = convert_to_zero[i - 1] + count
        # [i, n]
        for j in range(n - 1, -1, -1):
            count = (0 if s[j] == '1' else 1)
            convert_to_one[j] = convert_to_one[j + 1] + count
        res = float('inf')
        for x, y in zip(convert_to_zero, convert_to_one):
            res = min(res, x + y)
        return res
